### PR TITLE
EnvoyFilter: add ability to match on proxy metadata and version

### PR DESCRIPTION
**Pull request by rshriram**
Based on @ramaraochavali's concerns that EnvoyFilters would pose
a potential problem during upgrades, adding a proxy version check
in the match condition. This will allow users to apply the patches
to specific versions of the proxy

Signed-off-by: Shriram Rajagopalan <rshriram@tetrate.io>

### CODE REVIEW

**Comment by rshriram** [comment]
@ramaraochavali FYI. This should hopefully alleviate the upgrade issue. We can add more checks under the Proxy Version over time.

File: networking/v1alpha3/envoy_filter.proto [hunk 1]
```diff
@@ -366,6 +366,23 @@ message EnvoyFilter {
...
 
+  // ProxyVersion restricts the patches to sidecars or gateways with
+  // from a specific version of Istio. The Istio version for a given
+  // proxy is obtained from the node metadata field ISTIO_VERSION
+  // supplied by the proxy when connecting to Pilot. This value is
+  // embedded as an environment variable (ISTIO_META_ISTIO_VERSION) in
+  // the Istio proxy docker image. Custom proxy implementations should
+  // provide this metadata variable to take advantage of the Istio
+  // version check option.
+  message ProxyVersion {
```
Comment (ramaraochavali) [inline networking/v1alpha3/envoy_filter.proto]: @rshriram This is great. Thanks for handling this.

Comment (rshriram) [reply networking/v1alpha3/envoy_filter.proto]: actually check the proto again. I have made this a lot more generic.

**Comment by howardjohn** [comment]
This is not OWNERs approved by the way. I think there is a bug in our branch protection

File: networking/v1alpha3/envoy_filter.proto [hunk 2]
```diff
@@ -366,6 +366,35 @@ message EnvoyFilter {
     GATEWAY = 3;
   };
 
+  // One or more properties of the proxy to match on.
+  message ProxyMatch {
+    oneof version_match {
```
Comment (louiscryan) [inline networking/v1alpha3/envoy_filter.proto]: why a oneof ? why not a regex ?

**Comment by howardjohn** [comment]
This was not approved by owners @rshriram ..

File: networking/v1alpha3/envoy_filter.proto (hunk 2 above)
Comment (rshriram) [reply networking/v1alpha3/envoy_filter.proto]: I dont mind doing regex. But i thought you were strongly opposed to doing regexes at all. :)

**Comment by rshriram** [comment]
I talked to @louiscryan offline. So no harm done..

**Pull request merged**